frame/system/src/lib.rs

```diff
@@ -1194,7 +1200,10 @@ impl<T: Trait + Send + Sync> CheckWeight<T> where
 		let current_weight = Module::<T>::all_extrinsics_weight();
 		let maximum_weight = T::MaximumBlockWeight::get();
 		let limit = Self::get_dispatch_limit_ratio(info.class) * maximum_weight;
-		let added_weight = info.weight.min(limit);
+		let extrinsic_weight = info.weight.checked_add(T::ExtrinsicBaseWeight::get()).ok_or(InvalidTransaction::ExhaustsResources)?;
+		// TODO: Verify this line is sane, seems to be bad.
+		let added_weight = extrinsic_weight.min(limit);
+		// TODO: Should be checked add I think.
```

Comment:
I totally agree with this. end of the day would be great to see a test in which a transaction with very high weight in `custom::decl_module` to fail even as the first transaction.

This logic here was only correct for 
- chains with fixed weights
- assuming that the weight of no single individual transaction is more than the limit (basically the `min` was defensive and should have never happened. )

Reply:
I have updated the logic, please have a look :)